hash.go

```diff
@@ -407,9 +453,11 @@ func (h *sha224Hash) Sum(in []byte) []byte {
 
 // NewSHA256 returns a new SHA256 hash.
 func NewSHA256() hash.Hash {
-	return &sha256Hash{
-		evpHash: newEvpHash(crypto.SHA256, 256/8, 64),
+	h := newEvpHash(crypto.SHA256, 256/8, 64)
+	if isHashMarshallable(crypto.SHA256) {
+		return &sha256Marshal{sha256Hash{evpHash: h}}
 	}
+	return &sha256Hash{evpHash: h}
```

Comment:
It looks like a tiny bit more in common could be pulled out:

```go
	h := sha256Hash{evpHash: newEvpHash(crypto.SHA256, 256/8, 64)}
	if isHashMarshallable(crypto.SHA256) {
		return &sha256Marshal{h}
	}
	return &h
```

(Unless there's a perf/rule-of-thumb reason not to?)

Reply:
Good shot. I like it 😸